Source/Dafny/Compiler-cpp.cs

```diff
@@ -0,0 +1,2338 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright (C) Amazon.  All Rights Reserved.
+//
+//-----------------------------------------------------------------------------
+using System;
+using System.CodeDom;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.IO;
+using System.Diagnostics.Contracts;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using Bpl = Microsoft.Boogie;
+
+namespace Microsoft.Dafny {
+  public class CppCompiler : Compiler {
+    public CppCompiler(ErrorReporter reporter, ReadOnlyCollection<string> otherHeaders)
+    : base(reporter) {
+      this.headers = otherHeaders;
+      this.datatypeDecls = new List<DatatypeDecl>();
+      this.classDefaults = new List<string>();
+    }
+
+    private ReadOnlyCollection<string> headers;
+    private List<DatatypeDecl> datatypeDecls;
+    private List<string> classDefaults;
+    
+    /*
+     * Unlike other Dafny and Dafny's other backends, C++ cares about
+     * the order in which types are declared.  To make this more likely
+     * to succeed, we emit type information as gradually as possible
+     * in hopes that definitions are in place when needed.
+     */
+
+    // Forward declarations of class and struct names
+    private TargetWriter modDeclsWr = null;
+    private TargetWriter modDeclWr = null;
+    // Dafny datatype declarations
+    private TargetWriter dtDeclsWr = null;
+    private TargetWriter dtDeclWr = null;
+    // Dafny class declarations
+    private TargetWriter classDeclsWr = null;
+    private TargetWriter classDeclWr = null;
+    // Dedicated hash-function definitions for each type
+    private TargetWriter hashWr = null;
+
+    // Shadowing variables in Compiler.cs
+    new string DafnySetClass = "DafnySet";
+    new string DafnyMultiSetClass = "DafnyMultiset";
+    new string DafnySeqClass = "DafnySequence";
+    new string DafnyMapClass = "DafnyMap";
```

Comment:
Override them instead

Reply:
I don't think C# will let me override a field, will it?  Or are you suggesting upgrading the existing fields in `Compiler.cs` to properties so that they can be overridden?

Reply:
Yes, I was thinking converting the original types to be virtual property getters. I just suggested the same thing in https://github.com/dafny-lang/dafny/pull/696, where this also came up. We can adjust these once these PRs are both merged into master.

Reply:
Okay, sounds good to me.